lib/spack/spack/build_systems/cached_cmake.py

```diff
@@ -162,7 +162,9 @@ def initconfig_compiler_entries(self):
             ld_flags = " ".join(flags["ldflags"])
             ld_format_string = "CMAKE_{0}_LINKER_FLAGS"
             # CMake has separate linker arguments for types of builds.
-            for ld_type in ["EXE", "MODULE", "SHARED", "STATIC"]:
+            # 'ldflags' should not be used with CMAKE_STATIC_LINKER_FLAGS which
+            # is used by the archiver, so don't include "STATIC" in this loop:
+            for ld_type in ["EXE", "MODULE", "SHARED"]:
```

Comment:
I'm curious if @white238 agrees with this

Reply:
For some context: I ran into this issue when building RAJA and I had `ldflags` setup to be `-Wl,--undefined-version`. Then I got a failure in the following archive command from cmake:
```console
/usr/bin/llvm-ar qc lib/libRAJA.a -Wl,--undefined-version CMakeFiles/RAJA.dir/src/AlignedRangeIndexSetBuilders.cpp.o ...
```
After some digging, I found that `CMAKE_STATIC_LINKER_FLAGS` was set to `-Wl,--undefined-version` which was causing it to appear in the above command. From there, I tracked the place where `CMAKE_STATIC_LINKER_FLAGS` was set to the above code.

Reply:
I would agree @v-dobrev is correct in this fix and this CMake page backs that up.

https://cmake.org/cmake/help/latest/envvar/LDFLAGS.html

Though this opens a different question, that if we want to set a global `CMAKE_STATIC_LINKER_FLAGS`, I am not sure how we would do that with Spack's compiler flags. @becker33, thoughts?

Reply:
For flags that we want to apply for `ar`

* we could introduce another `*flags` property (alongside `cflags` etc.)
* or we could add logic to filter out all flags that are not compatible with `ar` from `ldflags` (I think the first suggestion is more clear though)